rush-plugins/rush-resolver-cache-plugin/src/afterInstallAsync.ts

```diff
@@ -0,0 +1,357 @@
+
+  terminal.writeLine(`Resolver cache will be written at ${cacheFilePath}`);
+
+  const contexts: Map<string, IResolverContext> = new Map();
+  const missingOptionalDependencies: Set<string> = new Set();
+
+  const projectByImporterPath: LookupByPath<RushConfigurationProject> =
+    rushConfiguration.getProjectLookupForRoot(workspaceRoot);
+
+  // Acquiring the libc version is a bit more obnoxious than platform and arch,
+  // but all of them are ultimately on the same object.
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  const { platform, arch, glibcVersionRuntime } = (process.report?.getReport() as any)?.header ?? process;
+  const libc: string = glibcVersionRuntime ? 'glibc' : 'musl';
+
+  // Enumerate external dependencies first, to simplify looping over them for store data
+  for (const [key, pack] of shrinkwrapFile.packages) {
+    let name: string | undefined = pack.name;
+    const descriptionFileRoot: string = getDescriptionFileRootFromKey(key, name);
+
+    // Skip optional dependencies that are incompatible with the current environment
+    if (pack.optional) {
+      if (pack.os && !pack.os.some((value) => value.toLowerCase() === platform)) {
+        missingOptionalDependencies.add(descriptionFileRoot);
+        continue;
+      }
+      if (pack.cpu && !pack.cpu.some((value) => value.toLowerCase() === arch)) {
+        missingOptionalDependencies.add(descriptionFileRoot);
+        continue;
+      }
+      if (pack.libc && !pack.libc.some((value) => value.toLowerCase() === libc)) {
+        missingOptionalDependencies.add(descriptionFileRoot);
+        continue;
+      }
+    }
+
+    const integrity: string | undefined = pack.resolution?.integrity;
+    if (!integrity) {
+      throw new Error(`Missing integrity for ${key}`);
+    }
+
+    if (!name && key.startsWith('/')) {
+      const versionIndex: number = key.indexOf('@', 2);
+      name = key.slice(1, versionIndex);
+    }
+
+    if (!name) {
+      throw new Error(`Missing name for ${key}`);
+    }
+
+    const context: IResolverContext = {
+      descriptionFileRoot,
+      descriptionFileHash: integrity,
+      isProject: false,
+      name,
+      deps: new Map(),
+      ordinal: contexts.size,
+      optional: pack.optional
+    };
+
+    contexts.set(descriptionFileRoot, context);
+
+    if (pack.dependencies) {
+      resolveDependencies(pack.dependencies, context);
+    }
+    if (pack.optionalDependencies) {
+      resolveDependencies(pack.optionalDependencies, context);
+    }
+  }
+
+  // For external packages, update the contexts with data from the pnpm store
+  // This gives us the list of nested package.json files, as well as the actual package name
+  // We could also cache package.json contents, but that proves to be inefficient.
+  await Async.forEachAsync(
+    contexts.values(),
+    async (context: IResolverContext) => {
+      const { descriptionFileRoot, descriptionFileHash } = context;
+      // Convert an integrith hash like
```

Comment:
There isn't a utility function from pnpm that does this?

Reply:
Eh. Not one easily accessible.